bittensor/_dataset/dataset_impl.py

```diff
@@ -1,538 +1,494 @@
+            no_tokenizer: bool = False,
+            data_dir: str =  os.path.expanduser('~/./bittensor/data'),
+            save_dataset : bool = False,
+            load_dataset : bool = True,
+            run_generator:bool=False,
+            buffer_size:int=100,
+            num_batches: int = 100,
+            max_datasets: int = 2,
+            num_folders: int = 10,
+            max_directories: int=10,
+            cache_size: int = 10, 
+            cache_calls_per_block: int=100):
+
+        """
+        Args:
+            batch_size (int):
+                The size of the batch.
+            sequence_length (int):
+                The length of the seqeunce (tokenwise).
+            block_size (int): 
+                The  size of the text data block which is used to create 
+                multiple samples.
+            num_workers (int):
+                Number of workers for pytorch Dataset.
+            datasets (Optional[Union[List[str], str]]):
+                List of dataset names to include from the pile.
+            loop ('asyncio.loop'):
+                Asyncio loop for class, defaults to default event loop.
+            no_tokenizer (bool):
+                Do not inlcude tokenizer.
+            data_dir (str):
+                Directory for saving assets.
+            save_dataset (bool):
+                Save the dataset text hashes.
+            load_dataset (bool):
+                Load the dataset text hashes.
+            run_generator (bool):
+                Run the background loop for fecthing data blocks.
+            buffer_size (int):
+                Size of the queue for asynchronously putting data blocks.
+            num_batches (int):
+                Number of generated batches in epoch.
+            max_datasets (int):
+                Max number of datasets.
+            cache_size (int):
+                size of blocks to cache while training
+            cache_calls_per_block (int):
+                calls per block when caching
+
+        """
+        self.__infinite_dataset_iterator = None
+        self.dataset_size = 0
         self.batch_size = batch_size
+        self.block_size = block_size
         self.num_workers = num_workers
-        self.tokenizer = bittensor.tokenizer( version = bittensor.__version__ )
-        self.dataset_name = dataset_name
-        self.data_dir = data_dir
-        self.save_dataset = save_dataset
-        self.datafile_size_bound = 262158
+        self.sequence_length = sequence_length
+        self.datasets = datasets
+        self.num_folders = num_folders
+        self.set_event_loop(loop=loop)
+        # if datasets is None then refer to all of the availabe datasets 
         self.max_datasets = max_datasets
-        self.__infinite_dataset_iterator = None
+        assert len(self.datasets) > 0
+        if self.datasets == 'default':
+            self.datasets = self.available_datasets
+        self.datasets = self.datasets[:self.max_datasets]
```

Comment:
`datasets` is of type `Optional[Union[List[str], str]]`. As an end-user I might understandably read this as if I can pass in just `"ArXiv"` (because passing in a list for a single dataset is counterintuitive). We should either support this behaviour, or not allow strings at all. (and have an empty list or `None` be the default value and behave as "default" does now).

Reply:
I had left it to have default or a list, but i see what you mean. I can have default and None as the trigger for every dataset.